Services/Mail/classes/BackgroundTask/class.ilMassMailTaskProcessor.php

```diff
@@ -0,0 +1,179 @@
+use ILIAS\DI\Container;
+
+/**
+ * @author  Niels Theen <ntheen@databay.de>
+ */
+class ilMassMailTaskProcessor
+{
+	/**
+	 * @var TaskManager
+	 */
+	private $taskManager;
+
+	/**
+	 * @var TaskFactory
+	 */
+	private $taskFactory;
+
+	/**
+	 * @var ilLanguage|null
+	 */
+	private $language;
+
+	/**
+	 * @var ilLogger|null
+	 */
+	private $logger;
+
+	/**
+	 * @var ilMailValueObjectJsonService|null
+	 */
+	private $objectJsonService;
+
+	/**
+	 * @param TaskManager $taskManager
+	 * @param TaskFactory|null $taskFactory
+	 * @param ilLanguage|null $language
+	 * @param ilLogger|null $logger
+	 * @param Container|null $dic
+	 * @param ilMailValueObjectJsonService|null $objectJsonService
+	 */
+	public function __construct(
+		TaskManager $taskManager = null,
+		TaskFactory $taskFactory = null,
+		ilLanguage  $language = null,
+		ilLogger $logger = null,
+		Container $dic = null,
+		ilMailValueObjectJsonService $objectJsonService = null
+	) {
+
+		if (null === $dic) {
+			global $DIC;
+			$dic = $DIC;
+		}
+
+		if (null === $taskManager) {
+			$taskManager = $dic->backgroundTasks()->taskManager();
+		}
+		$this->taskManager = $taskManager;
+
+		if (null === $taskFactory) {
+			$taskFactory = $dic->backgroundTasks()->taskFactory();
+		}
+		$this->taskFactory = $taskFactory;
+
+		if (null === $language) {
+			$language = $dic->language();
+		}
+		$this->language = $language;
+
+		if (null === $logger) {
+			$logger = ilLoggerFactory::getLogger('mail');
+		}
+		$this->logger = $logger;
+
+		if (null === $objectJsonService) {
+			$objectJsonService = new ilMailValueObjectJsonService();
+		}
+		$this->objectJsonService = $objectJsonService;
+	}
+
+	/**
+	 * @param ilMailValueObject[] $mailValueObjects - One MailValueObject = One Task
+	 * @param int $userId - User ID of the user who executes the background task
+	 * @param string $contextId - context ID of the Background task
+	 * @param array $contextParameters - context parameters for the background tasks
+	 * @param int $mailsPerTask - Defines how many mails will be added before a background task is executed
+	 * @throws ilException
+	 */
+	public function run(
+		array $mailValueObjects,
+		int $userId,
+		string $contextId,
+		array $contextParameters,
+		int $mailsPerTask = 100
+	) {
+		$objectsServiceSize = sizeof($mailValueObjects);
+
+		if ($objectsServiceSize <= 0) {
+			throw new ilException('First parameter must contain at least 1 array element');
+		}
+
+		if ($mailsPerTask <= 0) {
+			throw new ilException(sprintf('The mails per task MUST be a positive integer, "%s" given', $mailsPerTask));
+		}
+
+		$lastTask = null;
+		$taskCounter = 0;
+
+		$remainingObjects = array();
+		foreach ($mailValueObjects as $mailValueObject) {
+			if (false === ($mailValueObject instanceof ilMailValueObject)) {
+				throw new ilException('Array MUST contain ilMailValueObjects ONLY');
+			}
+
+			$taskCounter++;
+
+			$remainingObjects[] = $mailValueObject;
+			if ($taskCounter === $mailsPerTask) {
```

Comment:
Maybe building chunks here has to be based on the complete size/strlen of the JSON string, compared to the maximum length of the `clob` database field type (which is used for a `new SingleType(StringValue::class)` type).